### PR TITLE
[InstCombine] Resolve TODO: nnan nsz X / -0.0 -> copysign(inf, X)

### PR DESCRIPTION
**Pull request by AZero13**

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 1]
```diff
@@ -1600,7 +1600,17 @@ Instruction *InstCombinerImpl::foldFDivConstantDivisor(BinaryOperator &I) {
   // nnan X / +0.0 -> copysign(inf, X)
   if (I.hasNoNaNs() && match(I.getOperand(1), m_Zero())) {
     IRBuilder<> B(&I);
-    // TODO: nnan nsz X / -0.0 -> copysign(inf, X)
+    CallInst *CopySign = B.CreateIntrinsic(
+        Intrinsic::copysign, {C->getType()},
+        {ConstantFP::getInfinity(I.getType()), I.getOperand(0)}, &I);
+    CopySign->takeName(&I);
+    return replaceInstUsesWith(I, CopySign);
+  } else if (I.hasNoNaNs() && I.hasNoSignedZeros() &&
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1608]: NO else after return

File: llvm/test/Transforms/InstCombine/fdiv.ll [hunk 2]
```diff
@@ -992,3 +992,22 @@ define float @fdiv_nnan_neg_zero_f32(float %x) {
...
+
+define double @test_positive_zero(double %X) {
+; CHECK-LABEL: @test_positive_zero(
+; CHECK-NEXT:    [[TMP1:%.*]] = call nnan nsz double @llvm.copysign.f64(double 0x7FF0000000000000, double [[X:%.*]])
+; CHECK-NEXT:    ret double [[TMP1]]
+;
+  %1 = fdiv nnan nsz double %X, 0.0
+  ret double %1
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fdiv.ll:1004]: Test missing individual flags cases, and vectors

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp (hunk 1 above)
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1608]: In a follow up patch, if you wanted to be more aggressive you could also check for known-non-nan sources

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 3]
```diff
@@ -1600,7 +1600,18 @@ Instruction *InstCombinerImpl::foldFDivConstantDivisor(BinaryOperator &I) {
...
+        Intrinsic::copysign, {C->getType()},
+        {ConstantFP::getInfinity(I.getType()), I.getOperand(0)}, &I);
+    CopySign->takeName(&I);
+    return replaceInstUsesWith(I, CopySign);
+  }
+
+  // nnan nsz X / -0.0 -> copysign(inf, X)
+  if (I.hasNoNaNs() && I.hasNoSignedZeros() &&
+      match(I.getOperand(1), m_Specific(ConstantFP::get(
+                                 Type::getDoubleTy(I.getContext()), -0.0)))) {
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1613]: Can use m_NegZeroFP

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 4]
```diff
@@ -1600,7 +1600,18 @@ Instruction *InstCombinerImpl::foldFDivConstantDivisor(BinaryOperator &I) {
   // nnan X / +0.0 -> copysign(inf, X)
   if (I.hasNoNaNs() && match(I.getOperand(1), m_Zero())) {
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1601]: Should this be m_PosZeroFP?

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 5]
```diff
@@ -1600,7 +1600,18 @@ Instruction *InstCombinerImpl::foldFDivConstantDivisor(BinaryOperator &I) {
...
-    // TODO: nnan nsz X / -0.0 -> copysign(inf, X)
+    CallInst *CopySign = B.CreateIntrinsic(
+        Intrinsic::copysign, {C->getType()},
+        {ConstantFP::getInfinity(I.getType()), I.getOperand(0)}, &I);
+    CopySign->takeName(&I);
+    return replaceInstUsesWith(I, CopySign);
+  }
+
+  // nnan nsz X / -0.0 -> copysign(inf, X)
+  if (I.hasNoNaNs() && I.hasNoSignedZeros() &&
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1611]: I'm not seeing the difference between this case and the above one, they are both inserting copysign(inf, x)?

**Comment by AtariDreams** [comment]
@arsenm All issues addressed!

**Review by arsenm — COMMENTED** [review]
lgtm with some additional edge case handling

File: llvm/test/Transforms/InstCombine/fdiv.ll [hunk 6]
```diff
@@ -992,3 +992,22 @@ define float @fdiv_nnan_neg_zero_f32(float %x) {
...
+define double @test_negative_zero(double %X) {
+; CHECK-LABEL: @test_negative_zero(
+; CHECK-NEXT:    [[TMP1:%.*]] = call nnan nsz double @llvm.copysign.f64(double 0x7FF0000000000000, double [[X:%.*]])
+; CHECK-NEXT:    ret double [[TMP1]]
+;
+  %1 = fdiv nnan nsz double %X, -0.0
+  ret double %1
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fdiv.ll:1013]: I think some tests got lost since before? Should have with and without nsz, with + and -0, and a case with the mixed 0 vector

**Comment by AtariDreams** [comment]
Did, but now unrelated tests are failing, and the script seems unable to regenerate them. What do I do?

File: llvm/test/CodeGen/PowerPC/aix-small-local-exec-tls-largeaccess.ll [hunk 7]
```diff
@@ -227,4 +227,4 @@ entry:
 ; DIS:      000000000000be6c (idx: [[#NFA+19]]) mySmallLocalExecTLS3[TL]:
 ; DIS:      000000000000fcec (idx: [[#NFA+21]]) mySmallLocalExecTLS4[TL]:
 ; DIS:      0000000000013b6c (idx: [[#NFA+23]]) mySmallLocalExecTLS5[TL]:
-; DIS:      00000000000179ec (idx: [[#NFA+25]]) mySmallLocalExecTLSv2[TL]:
\ No newline at end of file
+; DIS:      00000000000179ec (idx: [[#NFA+25]]) mySmallLocalExecTLSv2[TL]:
```
Comment (dtcxzyw) [inline llvm/test/CodeGen/PowerPC/aix-small-local-exec-tls-largeaccess.ll:230]: These PowerPC changes are unrelated to your patch.

Comment (AtariDreams) [reply llvm/test/CodeGen/PowerPC/aix-small-local-exec-tls-largeaccess.ll:230]: No because they fail without this change

**Comment by AtariDreams** [comment]
@dtcxzyw these tests fail otherwise and I have no clue why

File: llvm/test/CodeGen/PowerPC/aix-small-local-exec-tls-largeaccess.ll (hunk 7 above)
Comment (dtcxzyw) [reply llvm/test/CodeGen/PowerPC/aix-small-local-exec-tls-largeaccess.ll:230]: https://github.com/llvm/llvm-project/pull/80162#issuecomment-1930037809

Comment (AtariDreams) [reply llvm/test/CodeGen/PowerPC/aix-small-local-exec-tls-largeaccess.ll:230]: Addressed!

File: llvm/test/Transforms/InstCombine/fdiv.ll (hunk 6 above)
Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/fdiv.ll:1013]: Addressed!

**Review by arsenm — APPROVED** [review]
